nixos/modules/services/web-apps/sftpgo.nix

```diff
@@ -0,0 +1,377 @@
+{ config, lib, pkgs, utils, ... }:
+
+with lib;
+
+let
+  cfg = config.services.sftpgo;
+  defaultUser = "sftpgo";
+  settingsFormat = pkgs.formats.json {};
+  configFile = settingsFormat.generate "sftpgo.json" cfg.settings;
+  hasPrivilegedPorts = any (port: port > 0 && port < 1024) (
+    catAttrs "port" (cfg.settings.httpd.bindings
+      ++ cfg.settings.ftpd.bindings
+      ++ cfg.settings.sftpd.bindings
+      ++ cfg.settings.webdavd.bindings
+    )
+  );
+in
+{
+  options.services.sftpgo = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = mdDoc "sftpgo";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.sftpgo;
+      defaultText = literalExpression "pkgs.sftpgo";
+      description = mdDoc ''
+        Which SFTPGo package to use.
+      '';
+    };
+
+    extraArgs = mkOption {
+      type = with types; listOf str;
+      default = [];
+      description = mdDoc ''
+        Additional command line arguments to pass to the sftpgo daemon.
+      '';
+      example = [ "--log-level" "info" ];
+    };
+
+    dataDir = mkOption {
```

Comment:
You use 3 different methods to create this directory: `StateDirectory`, `tmpfiles`, and `createHome`.

You should not use `createHome`. In my opinion you should use `StateDirectory`.

Reply:
`StateDirectory` only works if it is inside `/var/lib`.

I would argue that the dataDir should not be configurable.

Reply:
> `StateDirectory` only works if it is inside `/var/lib`.

Indeed. A good pattern in NixOS is to leave the user with an escape hatch that we don't provision directories for them. When dealing with network filesystems, etc... this can cause users problems. See an example of what is a good way to handle state directory creation here: https://github.com/NixOS/nixpkgs/blob/de5448dab588ad41aef40f8c7c0c230981656698/nixos/modules/services/security/cfssl.nix#L11


> maybe do `createHome` only if the dataDir is _not_ `/var/lib/sftpgo`?

Never use `createHome` to create state directories because it ends poorly for some users needs (see prior comment about network mounts).

> Some might argue that the dataDir should not be configurable in the first place, and it is the user's task to somehow bind-mound or symlink the data to /var/lib.

This certainly would make life easier for developers... though at the cost of the end user. Conditionally creating the directory as in the example I linked above is the best compromise on my opinion.